core/src/blocks/curl.rs

```diff
@@ -0,0 +1,239 @@
+    error: Option<String>,
+}
+
+#[async_trait]
+impl Block for Curl {
+    fn block_type(&self) -> BlockType {
+        BlockType::Curl
+    }
+
+    fn inner_hash(&self) -> String {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update("curl".as_bytes());
+        hasher.update(self.method.as_bytes());
+        hasher.update(self.url.as_bytes());
+        hasher.update(self.headers_code.as_bytes());
+        hasher.update(self.body_code.as_bytes());
+        format!("{}", hasher.finalize().to_hex())
+    }
+
+    async fn execute(&self, _name: &str, env: &Env) -> Result<Value> {
+        let method = match self.method.as_str() {
+            "GET" => Method::GET,
+            "POST" => Method::POST,
+            "PUT" => Method::PUT,
+            "PATCH" => Method::PATCH,
+            _ => Err(anyhow!(
+                "Invalid method {}, supported methods are GET, POST, PUT.",
+                self.method
+            ))?,
+        };
+
+        let e = env.clone();
+        let headers_code = self.headers_code.clone();
+        let headers_value: Value = match tokio::task::spawn_blocking(move || {
+            let mut script = Script::from_string(headers_code.as_str())?
+                .with_timeout(std::time::Duration::from_secs(10));
+            script.call("_fun", (&e,))
+        })
+        .await?
+        {
+            Ok(v) => v,
+            Err(e) => Err(anyhow!("Error in headers code: {}", e))?,
+        };
+
+        let e = env.clone();
+        let body_code = self.body_code.clone();
+        let body_value: Value = match tokio::task::spawn_blocking(move || {
+            let mut script = Script::from_string(body_code.as_str())?
+                .with_timeout(std::time::Duration::from_secs(10));
+            script.call("_fun", (&e,))
+        })
+        .await?
+        {
+            Ok(v) => v,
+            Err(e) => Err(anyhow!("Error in body code: {}", e))?,
+        };
+
+        let url = replace_variables_in_string(&self.url, "url", env)?;
+        let parsed_url = Url::parse(url.as_str())?;
+
+        match parsed_url.scheme() {
+            "https" => (),
+            _ => Err(anyhow!("Only the `https` scheme is authorized."))?,
+        }
+
+        let _: Vec<std::net::Ipv4Addr> = match parsed_url.host() {
+            Some(h) => match h {
+                Host::Domain(d) => {
+                    let ipv4: Vec<std::net::Ipv4Addr> = lookup_host(d)?
+                        .into_iter()
+                        .filter_map(|ip| match ip {
+                            std::net::IpAddr::V4(ip) => Some(ip),
+                            _ => None,
+                        })
+                        .collect::<Vec<_>>();
+                    match ipv4.len() {
+                        0 => Err(anyhow!("Could not find an ipv4 address for host: {}", d))?,
+                        _ => ipv4,
+                    }
+                }
+                Host::Ipv4(ip) => vec![ip],
```

Comment:
I would strongly suggest erroring out on any manually entered IP address. Anything you can do to lower the attack surface here seems worth it, and the number of legitimate services that are at IP addresses only is vanishingly small.

Reply:
Unless you want to hit some of your own service you scrapily deployed?

Reply:
Even folks who are scrappily doing things usually have a FQDN these days, like a *.repl.co URL, or an *.aws.amazon.com URL. I really don't think it's worth the risk.

Reply:
it's very rare you hit a service IP, often you have some generated subdomains